internal/provider/common_key.go

```diff
@@ -0,0 +1,190 @@
+var keyGenerators = map[Algorithm]keyGenerator{
+	RSA: func(d *schema.ResourceData) (crypto.PrivateKey, error) {
+		rsaBits := d.Get("rsa_bits").(int)
+		return rsa.GenerateKey(rand.Reader, rsaBits)
+	},
+	ECDSA: func(d *schema.ResourceData) (crypto.PrivateKey, error) {
+		curve := ECDSACurve(d.Get("ecdsa_curve").(string))
+		switch curve {
+		case P224:
+			return ecdsa.GenerateKey(elliptic.P224(), rand.Reader)
+		case P256:
+			return ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+		case P384:
+			return ecdsa.GenerateKey(elliptic.P384(), rand.Reader)
+		case P521:
+			return ecdsa.GenerateKey(elliptic.P521(), rand.Reader)
+		default:
+			return nil, fmt.Errorf("invalid ECDSA curve; supported values are: %v", SupportedECDSACurves())
+		}
+	},
+	ED25519: func(d *schema.ResourceData) (crypto.PrivateKey, error) {
+		_, key, err := ed25519.GenerateKey(rand.Reader)
+		if err != nil {
+			return nil, fmt.Errorf("failed to generate ED25519 key: %s", err)
+		}
+		return &key, err
+	},
+}
+
+// keyParsers provides a keyParser given a specific Algorithm.
+var keyParsers = map[Algorithm]keyParser{
+	RSA: func(der []byte) (crypto.PrivateKey, error) {
+		return x509.ParsePKCS1PrivateKey(der)
+	},
+	ECDSA: func(der []byte) (crypto.PrivateKey, error) {
+		return x509.ParseECPrivateKey(der)
+	},
+	ED25519: func(der []byte) (crypto.PrivateKey, error) {
+		return x509.ParsePKCS8PrivateKey(der)
+	},
+}
+
+// parsePrivateKeyPEM takes a slide of bytes containing a private key
+// encoded in [PEM (RFC 1421)](https://datatracker.ietf.org/doc/html/rfc1421) format,
+// and returns a crypto.PrivateKey implementation, together with the Algorithm used by the key.
+func parsePrivateKeyPEM(keyPEMBytes []byte) (crypto.PrivateKey, Algorithm, error) {
+	pemBlock, rest := pem.Decode(keyPEMBytes)
+	if pemBlock == nil {
+		return nil, "", fmt.Errorf("failed to decode PEM block: decoded bytes %d, undecoded %d", len(keyPEMBytes)-len(rest), len(rest))
+	}
+
+	// Map PEM Preamble of the Private Key to the corresponding Algorithm
+	algorithm, err := PEMPreamblePrivateKey(pemBlock.Type).Algorithm()
+	if err != nil {
+		return nil, "", err
+	}
+
+	// Identify parser for the given key, using the Algorithm
+	parser, ok := keyParsers[algorithm]
+	if !ok {
+		return nil, "", fmt.Errorf("unsupported key algorithm: %s", algorithm)
+	}
+
+	// Parse the specific crypto.PrivateKey from the PEM Block bytes
+	prvKey, err := parser(pemBlock.Bytes)
+	if err != nil {
+		return nil, "", fmt.Errorf("failed to parse private key of algorithm %s: %w", algorithm, err)
+	}
+
+	return prvKey, algorithm, nil
+}
+
+// parsePrivateKeyOpenSSHPEM takes a slide of bytes containing a private key
+// encoded in [OpenSSH PEM (RFC 4716)](https://datatracker.ietf.org/doc/html/rfc4716) format,
+// and returns a crypto.PrivateKey implementation, together with the Algorithm used by the key.
+func parsePrivateKeyOpenSSHPEM(keyOpenSSHPEMBytes []byte) (crypto.PrivateKey, Algorithm, error) {
```

Comment:
This is _net new_ in this file, and now the 1 function used to parse ALL OpenSSH PEM private keys across the provider.